src/Akeneo/Category/back/Infrastructure/Storage/Save/Query/PurgeOrphanCategoriesSql.php

```diff
@@ -0,0 +1,42 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Akeneo\Category\Infrastructure\Storage\Save\Query;
+
+use Doctrine\DBAL\Connection;
+
+/**
+ * Delete all orphan categories with their childs, those without parents and aren't tree (id = root).
+ *
+ * @copyright 2023 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class PurgeOrphanCategoriesSql implements PurgeOrphanCategories
+{
+    public function __construct(
+        private Connection $connection,
+    ) {
+    }
+
+    public function execute(): int
+    {
+        $sql = <<< SQL
+            WITH RECURSIVE orphan_categories as (
+                SELECT id, code, parent_id
+                FROM pim_catalog_category AS children
+                WHERE parent_id IS NULL AND root != id
+                UNION
+                SELECT pim_catalog_category.id, pim_catalog_category.code, pim_catalog_category.parent_id
+                FROM pim_catalog_category
+                INNER JOIN orphan_categories parent ON parent.id = pim_catalog_category.parent_id
+            )
+            DELETE
+            FROM pim_catalog_category
+            USING pim_catalog_category
+            JOIN orphan_categories ON pim_catalog_category.id = orphan_categories.id;
```

Comment:
I prefer to specify INNER, even if it is the default for JOIN in MySql
```suggestion
            INNER JOIN orphan_categories ON pim_catalog_category.id = orphan_categories.id;
```

Reply:
Good for me.

Reply:
Good for me to be more explicit.